transports/quic/src/transport.rs

```diff
@@ -0,0 +1,704 @@
+}
+
+impl<P> GenTransport<P> {
+    pub fn new(config: Config) -> Self {
+        let handshake_timeout = config.handshake_timeout;
+        let quinn_config = config.into();
+        Self {
+            listeners: SelectAll::new(),
+            quinn_config,
+            handshake_timeout,
+            dialer: HashMap::new(),
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<P: Provider> Transport for GenTransport<P> {
+    type Output = (PeerId, Muxer);
+    type Error = Error;
+    type ListenerUpgrade = Connecting;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn listen_on(&mut self, addr: Multiaddr) -> Result<ListenerId, TransportError<Self::Error>> {
+        let socket_addr =
+            multiaddr_to_socketaddr(&addr).ok_or(TransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new::<P>(
+            listener_id,
+            socket_addr,
+            self.quinn_config.clone(),
+            self.handshake_timeout,
+        )?;
+        self.listeners.push(listener);
+
+        // Remove dialer endpoint so that the endpoint is dropped once the last
+        // connection that uses it is closed.
+        // New outbound connections will use a bidirectional (listener) endpoint.
+        self.dialer.remove(&socket_addr.ip().into());
+
+        Ok(listener_id)
+    }
+
+    fn remove_listener(&mut self, id: ListenerId) -> bool {
+        if let Some(listener) = self.listeners.iter_mut().find(|l| l.listener_id == id) {
+            // Close the listener, which will eventually finish its stream.
+            // `SelectAll` removes streams once they are finished.
+            listener.close(Ok(()));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn address_translation(&self, listen: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        if !is_quic_addr(listen) || !is_quic_addr(observed) {
+            return None;
+        }
+        Some(observed.clone())
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| TransportError::MultiaddrNotSupported(addr.clone()))?;
+        if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        }
+        let mut listeners = self
+            .listeners
+            .iter_mut()
+            .filter(|l| {
+                if l.is_closed {
+                    return false;
+                }
+                let listen_addr = l.endpoint_channel.socket_addr();
+                SocketFamily::is_same(&listen_addr.ip(), &socket_addr.ip())
+                    && listen_addr.ip().is_loopback() == socket_addr.ip().is_loopback()
+            })
+            .collect::<Vec<_>>();
+
+        let dialing = if !listeners.is_empty() {
```

Comment:
Minor nit, but we may want to fast path the (probably?) common case of a single listener.

Reply:
Done in 6af111627576954b289805bf40657940305e442c. Wdyt?